core/src/databases/database.rs

```diff
@@ -79,9 +86,40 @@ impl Database {
         rows: Vec<DatabaseRow>,
         truncate: bool,
     ) -> Result<()> {
-        // This will be used to update the schema incrementally once we store schemas. For now this
-        // is a way to validate the content of the rows (only primitive types).
-        let _ = TableSchema::from_rows(&rows)?;
+        let table = match store
+            .load_database_table(
+                &self.project,
+                &self.data_source_id,
+                &self.database_id,
+                table_id,
+            )
+            .await?
+        {
+            Some(t) => t,
+            None => Err(anyhow!(
+                "Table {} not found in database {}",
+                table_id,
+                self.database_id
+            ))?,
+        };
+
+        let new_rows_table_schema = TableSchema::from_rows(&rows)?;
+        let table_schema = match table.schema() {
+            // If there is no existing schema cache, simply use the new schema.
+            None => new_rows_table_schema,
+            Some(existing_table_schema) => {
+                // If there is an existing schema cache, merge it with the new schema.
+                existing_table_schema.merge(&new_rows_table_schema)?
+            }
+        };
+
+        let update_table_schema_future = store.update_database_table_schema(
```

Comment:
? if it's just a clone on a small variable will make the code much easier to read

Reply:
> and then execute

It's executed right away, I'm just not awaiting it right away. 
This is basically like a await Promise.all() of this + batch_upsert_database_rows